test/iterator-test.cc

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) Google LLC.
+// SPDX-License-Identifier: MIT Licence
+//
+#include "fmt/iterator.h"
+
+#include <algorithm>
+#include <deque>
+#include <iterator>
+#include <vector>
+
+#include "fmt/ranges.h"
+#include "gtest-extra.h"
+#include "gtest/gtest.h"
+
+namespace {
+#if defined(__cpp_lib_ranges) and __cpp_lib_ranges >= 201911
+template <typename I> void check_iterator_conformance() {
+  static_assert(not std::readable<I>);
+  static_assert(std::output_iterator<I, int>);
+  static_assert(std::output_iterator<I, double>);
+  static_assert(std::output_iterator<I, std::string>);
+  static_assert(not std::output_iterator<I, std::vector<int>>);
+}
+
+template [[maybe_unused]] void
+check_iterator_conformance<fmt::print_iterator>();
+template [[maybe_unused]] void
+check_iterator_conformance<fmt::wprint_iterator>();
+template [[maybe_unused]] void
+check_iterator_conformance<fmt::u8print_iterator>();
+template [[maybe_unused]] void
+check_iterator_conformance<fmt::u16print_iterator>();
+template [[maybe_unused]] void
+check_iterator_conformance<fmt::u32print_iterator>();
+#endif  // __cpp_lib_ranges
+
+template <typename I> void check_default_operations() {
+  auto i = I{};
+  {
+    auto& prefix = ++i;
+    EXPECT_EQ(std::addressof(i), std::addressof(prefix));
+  }
+  {
+    auto& postfix = i++;
+    EXPECT_EQ(std::addressof(i), std::addressof(postfix));
+  }
+  {
+    auto& deref = *i;
+    EXPECT_EQ(std::addressof(i), std::addressof(deref));
+  }
+}
+
+TEST(PrintIterator, DefaultOperators) {
+  check_default_operations<fmt::print_iterator>();
+  check_default_operations<fmt::wprint_iterator>();
+  check_default_operations<fmt::u16print_iterator>();
+  check_default_operations<fmt::u32print_iterator>();
+#ifdef __cpp_char8_t
+  check_default_operations<fmt::u8print_iterator>();
+#endif  // __cpp_char8_t
+}
+
+enum class decorator { off, on };
+
+template <decorator is_decorated>
+void check_assignment(fmt::print_iterator simple_writer, std::FILE* out);
```

Comment:
Why not replace template with two separate functions?

Reply:
Yeah, that's the smarter option :sweat_smile: